### PR TITLE
Combine inferred shape and existing shape

### PR DESCRIPTION
**Pull request by tungld**
Previous patches #1654 and #1659 implemented a new shape inference feature that compares an inferred shape and an existing shape to choose the better one. Only unary element-wise ops were updated to use the new feature.

This patch enhances the feature by combining the inferred and existing shapes to produce a new shape, e.g. inferred shape `?x3` and existing shape is `2x?`, then we know that the final shape could be `2x3`.

In this patch, all ONNX and ZHigh ops are updated with the new shape inference feature.

While comparing the inferred and existing shapes, I found some wrong output shapes in lit tests for SpaceToDepth and Compress, and I fixed them.

### CODE REVIEW

File: src/Dialect/ONNX/ShapeInference/ONNXShapeHelper.cpp [hunk 1]
```diff
@@ -399,44 +400,57 @@ LogicalResult ONNXGenericPoolShapeHelper<OP_TYPE, OP_ADAPTOR>::computeShape(
...
+    if ((existingShape[i] != -1) && (inferredShape[i] == -1))
+      inferredShape[i] = existingShape[i];
     // inferedDim is different from existingDim. Believe in existingDim.
-    if ((existingDim != -1) && (inferedDim != -1) &&
-        (existingDim != inferedDim))
-      return false;
+    if ((existingShape[i] != -1) && (inferredShape[i] != -1) &&
+        (existingShape[i] != inferredShape[i])) {
+      // Warning for users.
+      llvm::outs() << "Shape inference: the inferred dim (" << inferredShape[i]
```
Comment (AlexandreEichenberger) [inline src/Dialect/ONNX/ShapeInference/ONNXShapeHelper.cpp:426]: isn't there an error stream? It would also be good to preface it with "Warning:"

**Review by AlexandreEichenberger — APPROVED** [review]
LGTM, I approved it but you may consider creating some helper functions to reduce the redundant code that is added all over to cover this new useful optimization on shapes.

File: src/Dialect/ONNX/ONNXOps.cpp [hunk 2]
```diff
@@ -4061,14 +4109,18 @@ LogicalResult ONNXEyeLikeOp::inferShapes(
   } else {
     elementType = inputType.getElementType();
   }
-  getResult().setType(RankedTensorType::get(inputType.getShape(), elementType));
+
```
[the remark this answers is not in the document]
Comment (tungld) [reply src/Dialect/ONNX/ONNXOps.cpp:4077]: Good suggestion. I created a helper function `updateType(tensor, shape, elementType, encoding)` to update the type of a tensor and put our optimization inside it. Thanks!

**Comment by tungld** [comment]
@jenkins-droid test this please

**Pull request merged**